src/main/scala/is/hail/methods/LDMatrix.scala

```diff
@@ -1,53 +1,134 @@
 
-    val normalizedIRM = ToNormalizedIndexedRowMatrix(vds)
-    normalizedIRM.rows.persist()
-    val variantKeptIndices = normalizedIRM.rows.map{case IndexedRow(idx, _) => idx.toInt}.collect()
-    assert(variantKeptIndices.isSorted, "Array of kept variants is not sorted. This is a bug.")
-    val variantsKept = variantKeptIndices.map(idx => originalVariants(idx))
+        // These ought to be sorted by variant number from time of construction of grouped RDD.
+        // Since they're sorted, it is safe to build matrix where entry i, j is computeR(g1Vars(i), g2Vars(j))
+        val g1Variants = groupMap.get(group1).get
+        val g2Variants = groupMap.get(group2).get
 
-    val nVariantsKept = variantsKept.length
-    val nVariantsDropped = originalVariants.length - nVariantsKept
+        val dataLength = g1Variants.length * g2Variants.length
+        val data = new Array[Double](dataLength)
 
-    info(s"Computing LD Matrix with ${variantsKept.length} variants using $nSamples samples. $nVariantsDropped variants were dropped.")
+        var i = 0
+        while (i < g2Variants.length) {
+          var j = 0
+          while (j < g1Variants.length) {
+            data(j + i * g2Variants.length) = LDPrune.computeR(g1Variants(j), g2Variants(i))
+            j += 1
+          }
+          i += 1
+        }
 
-    //The indices can be expected to be correct from the zip since the VDS is backed by an OrderedRDD of variants.
-    val normalizedFilteredRows = normalizedIRM.rows.zipWithIndex()
-      .map {case (IndexedRow(_, data), idx) => IndexedRow(idx, data)}
+        ((group1, group2), new DenseMatrix(g1Variants.length, g2Variants.length, data))
+      }
+    }
 
-    val normalizedBlockMatrix = new IndexedRowMatrix(normalizedFilteredRows).toBlockMatrixDense()
-    normalizedBlockMatrix.persist(StorageLevel.MEMORY_AND_DISK)
-    normalizedBlockMatrix.blocks.count()
-    normalizedIRM.rows.unpersist()
+    val allBlocks = computedBlocks.flatMap { case ((i, j), mat) =>
+      if (i == j) List(((i, j), mat)) else List(((i, j), mat), ((j, i), mat.transpose))
+    }
 
-    val indexedRowMatrix = (normalizedBlockMatrix multiply normalizedBlockMatrix.transpose)
-      .toIndexedRowMatrix()
+    val blockMatrix: BlockMatrix = new BlockMatrix(allBlocks, trueBlockSize, trueBlockSize)
+    val irm = blockMatrix.toIndexedRowMatrix()
 
-    val nSamplesInverse = 1.0 / nSamples
-    val scaledIndexedRowMatrix = new IndexedRowMatrix(indexedRowMatrix.rows.
-      map{case IndexedRow(idx, vals) => IndexedRow(idx, vals.map(d => d * nSamplesInverse))})
+    LDMatrix(irm, variantsKeptArray, nSamples)
+  }
+
+  /**
+    * Computes two maps. The first map specifies which partitions a given group is needed on. The second map
+    * specifies which blocks must be computed by a given partition.
+    * @param numberOfGroups The total number of groups.
+    * @param partitioner The partitioner used to divide up the blocks that need to be computed.
```

Comment:
I think having different terms for a block of variants and a block of matrix would be helpful. This may be why you had group in the first place, perhaps that's a good distinction after all. There is a block for each pair of groups of variants. (or maybe region or interval if variants are sorted)